community/bolt/src/main/java/org/neo4j/bolt/transport/pipeline/ProtocolHandshaker.java

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright (c) 2002-2018 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.bolt.transport.pipeline;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+
+import org.neo4j.bolt.BoltChannel;
+import org.neo4j.bolt.transport.BoltProtocolPipelineInstaller;
+import org.neo4j.bolt.transport.BoltProtocolPipelineInstallerFactory;
+import org.neo4j.logging.Log;
+import org.neo4j.logging.LogProvider;
+
+import static java.lang.String.format;
+
+public class ProtocolHandshaker extends ChannelInboundHandlerAdapter
```

Comment:
I think this class might benefit from extending `ReplayingDecoder`. It'll then won't have to bother with `handshakeBuffer` and reading exactly 5 ints from the channel.

Reply:
I checked the `ReplayingDecoder` but found it a bit more complex and cryptic. We'll either need to implement a re-entrant `#decode` method or add some states backed by an `Enum` to track where we left off. Furthermore, we'll still need to cater for left-over bytes in the buffer to send to the newly installed pipeline.

I didn't see much advantage over the current version - which is somewhat not that complex.